### ISSUE TITLE
Change for "Faster HTTP"

### ISSUE THREAD

**Issue by tracker1** [uWebSockets/uWebSockets#379]
Regarding faster http... why not have the arguments for `wss.onHttpRequest((socket, req) => {` match node's API, including the order (they already seem similar, but out of order)...

If they matched node's order and API, it would be easy enough to wire up express, koa or other options with your http implementation.

[1] https://github.com/uWebSockets/uWebSockets/wiki/WIP:-Faster-HTTP-for-Node.js

**Comment by pateketrueke** [comment]
IMO the arguments order is trivial, the whole object definition should be emulated or mapped someway: that's the key.

[I've been playing with it around a while](https://github.com/pateketrueke/homegrown/blob/master/examples/uws.js), very fast of course, but it need more work to be done.

I'm very interested on this, how can I contribute on this topic?

**Comment by alexhultman** [comment]
I wanted to implement Node.js's interface but I need Proxies to turn req.headers[] into req.getHeader(). Proxies are Node.js 7+ it seems and there are many differences other than this. The idea is to wrap things in uws to a minimal degree and then have adopters wrap it in something more Node.js ish. But I don't know yet, could change. I wouldn't bother with ExpressJs too much, there are other "web fra!eworks" being made every day. You just need a good product name and boom

**Comment by alexhultman** [comment]
@pateketrueke I think the best would be to keep playing with it and keep benchmarking it to make sure you are not regressing too much. Wrap things as much as possible and tell me what needs to change and be added and we can iterate on this. If you play with Proxies you should be able to get close to Node.js style.

**Comment by alexhultman** [comment]
I've done some thinking and I think the best idea is to simply try and wrap it close enough to Node.js style so that you can pass an uws http server module as parameter to the uws constructor. That could be the first big milestone - to be able to pass a native http server to uws itself and have it just work. I benchmarked routing the request handling through an EventEmitter a la Node.js and there was no real difference in performance and the http server doesn't have to be slim in memory. Will think some more but probably this is the only way to go forward. I think 150k requests/second is a good end goal.

**Comment by alexhultman** [comment]
I have a prototype here that looks like this:

```javascript
'use strict';

const uws = require('./dist/uws');

// note: this is uws's HTTP server, not Node.js's
const server = uws.http.createServer((req, res) => {
    res.end('Welcome to the 90s!\n');
});

// we can pass uws's own HTTP server to itself and have it run just like normally
const wss = new uws.Server({server: server});

// this part is like before, like ws
wss.on('connection', (ws) => {
    ws.send('Welcome to the 10s!');
});

// native listen
server.listen(3000);
```

This is kind of a pipe dream I had this morning that now is very close to working. And it does ~190k requests per second. uws.http.createServer returns a server object that "quacks" like Node.js's own http.createServer so it can be passed as argument to uws.Server and it just works.

**Comment by pateketrueke** [comment]
Interesting, let me try!

**Comment by alexhultman** [comment]
I pushed what I have to the pipedream branch. It is broken for upgrades currently. Lots of changes.

**Comment by japrescott** [comment]
just wanted to say that you guys (@alexhultman) are fucking crazy and incredibly awesome. I love reading about uws and the work you are up to. Keep up the great work!

**Comment by valentinvichnal** [comment]
Amazing, after finished can I use it to replace express's http.createServer?

**Comment by valentinvichnal** [comment]
@alexhultman Amazing, after finished can I use it to replace express's http.createServer?

**Comment by valentinvichnal** [comment]
@alexhultman Amazing, after finished can I use it to replace express's http.createServer, or it won't be fully compatible with the original?

**Comment by alexhultman** [comment]
I can't be 100% compatible but it can be "mostly compatible" just like the websocket support is mostly compatible with ws. The first step would be to get v0.13.0 out with a preview of the very basic HTTP server support so that people can play with it and give more feedback.

**Comment by ohenepee** [comment]
@alexhultman Which web framework woud you consider as fast?

**Comment by alexhultman** [comment]
Assuming we are talking about Node.js, the only "web framework" that doesn't want me to throw my screen out the window is https://github.com/aerojs/aero but it has a GPL license and is a dead project (no communication whatsoever via mail or issue tracker). But even this project has significant overhead.

Sails.js will make your mainframe data center perform like a 1$ calculator though, so Express.js doesn't look that bad when you compare (well it is terrible on its own).

TLDR; Don't use Node.js if you have any kind of soul.

**Comment by ronag** [comment]
koa is pretty decent. Use Node.js if you want get anything done in reasonable time. Another server is cheaper than developer time.

**Comment by ohenepee** [comment]
@ronag I use NodeJS indeed for time-related projects and of course expressiveness... but on the performance side, I just Go

**Comment by alexhultman** [comment]
I heard Ruby on Rails is what Google makes use of internally.

**Comment by ohenepee** [comment]
Well, there'll have to be a specific project/section you're referring to... cos Google use almost every language in their Search Engine

**Comment by alexhultman** [comment]
Yes the project I'm referring to is Project Sarcasm. It's a work in progress...

**Comment by ronag** [comment]
and has infinite funding...

**Comment by ronag** [comment]
Though in all seriousness... I've found that often you don't even need any framework like express or koa. Just use regular Nodejs HTTP api and pull in any functionality that you might need from npm.

**Comment by ohenepee** [comment]
Yes, but you still won't get that much performance (let's say about 3-4x increase compared to those frameworks).. whereas Go (Iris), Rust (Iron), and Python (Falcon) among others can give you 5-10x performance increase

**Comment by alexhultman** [comment]
Go is not "high performance". It is a GC:d language much like Java and C#. Rust, now this is something. Python is a fucking script, it is grammatically incorrect to put a script in the same sentence as "performance".

**Comment by alexhultman** [comment]
![](https://media.tenor.co/images/dca8120ae89cf5cb5c62ac088d28d5e4/tenor.gif)

**Comment by ronag** [comment]
Rust is quite beautiful.

**Comment by ohenepee** [comment]
@alexhultman By performance I meant "web" (HTTP) to be specific and yes Python and Go has some pretty good performance, something that Rust doesn't yet have.. the frameworks I listed have ultra-impressive performance. Check out https://www.techempower.com/benchmarks
